Enable `winsup` test suite compilation

thiru-mcw opened the pull request:
The internal test suite for Cygwin, located in the `winsup` directory, was previously removed from the build-system. This PR restores the test suite build process.

>After re-enabling it, some tests encountered compilation failures due to issues such as pointer type mismatches and old-style function declarations.

This was fixed by following commits in the upstream repository:

- [Cygwin: testsuite: Fix compilation for arm64](https://github.com/cygwin/cygwin/commit/6c5537c0298e9135ff27119c6a18dcc6f94459f5)
- [Cygwin: testsuite: Fix compatibility with GCC 15](https://github.com/cygwin/cygwin/commit/cce2ffd374e2ab4507cb973c74348cd1be4d106e)

This PR includes the necessary and workarounds to ensure the entire test suite builds and runs successfully.

thiru-mcw commented:
![image](https://github.com/user-attachments/assets/3c98a44f-0739-4041-9443-5a75524059b8)

Out of 26 skipped tests, 23 have been fixed and the fixes have been upstreamed. The remaining failures are as follows:

- `winsup.api/user_malloc` : Removed due to relocation issues (details in the table).
- `mingw/cygload` : Removed due to missing makefile rules (No rule to make target 'mingw/cygload').
- `winsup.api/ltp/symlink01`: Debugging is in-progress. The PR will be updated once fixed. [Ref: symlink01.log](https://github.com/user-attachments/files/19244740/error_log-symlink01.log)




Current status:
![image](https://github.com/user-attachments/assets/67733401-f8db-4309-a63a-58835a36c600)

**Edit:** `symlink01` test is also fixed, and changes are added to this PR

Blackhex commented on winsup/testsuite/Makefile.am:
```diff
@@ -322,15 +321,14 @@ LDADD_FOR_TESTDLL = $(builddir)/../cygwin/libcygwin.a -lgcc -lkernel32 -luser32
 # flags for test executables
 AM_CPPFLAGS = -I$(srcdir)/libltp/include
-AM_LDFLAGS = $(LDFLAGS_FOR_TESTDLL)
+AM_LDFLAGS = $(LDFLAGS_FOR_TESTDLL) -L$(builddir)/../../../../mingw/mingw-w64-crt/libarm64/
```
What library is missing without this? Can this be set using `LDFLAGS` in the build scripts? This would be preferred.

thiru-mcw replied:
The missing library was `-lwinmm`, which is required for building `devdsp test`.

while searching for `libwinmm.a` found 3 occurrences:

1. ./mingw-woarm64-build/build-aarch64-pc-cygwin-msvcrt/mingw-crt/libarm64/libwinmm.a
2. ./mingw-woarm64-build/build-aarch64-pc-cygwin-msvcrt/mingw/mingw-w64-crt/libarm64/libwinmm.a
3. ./toolchain_path/aarch64-pc-cygwin/lib/w32api/libwinmm.a

Adding `-L$(builddir)/../../../../mingw/mingw-w64-crt/libarm64/` allowed the linker to find the necessary library from the `./mingw-woarm64-build/build-aarch64-pc-cygwin-msvcrt/mingw/mingw-w64-crt/libarm64/` directory. Also I tried linking the same library from the other paths which also works in this case.

Blackhex replied:
From the options found, the `libwinmm.a` should be used from `$TOOLCHAIN_PATH/$TARGET/lib/w32api` directory rather than from the build folders - the build of Cygwin should not be dependent on the build of the toolchain. On the other hand, I don't know what is the reason why the dependency on this library is needed for `aarch64` while it's not the issue on `x64`? Does `x64` link with that library? From where it takes it? The relevant code adding `lib/w32api` folder to the linker paths seems to be https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/woarm64/configure.ac#L3227. I'd try to find the answers to the prevous questions there. The proper fix might also be more relevant to `configure.ac` than `Makefile.am`.

thiru-mcw replied:
> the libwinmm.a should be used from $TOOLCHAIN_PATH/$TARGET/lib/w32api directory 

@Blackhex Thanks for the info! I will implement the same if there is no other way.

>library is needed for aarch64 while it's not the issue on x64? Does x64 link with that library?

It is needed on x64 as well, it is directly linked during a native build (building Cygwin with the existing Cygwin runtime).

> I'd try to find the answers to the previous questions there.

https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/7a762b853e1a994616adab1e499c866629be0e74/winsup/testsuite/Makefile.am#L328-L329

I am currently not sure why this linking is not done in aarch64, will check and get back here.

eukarpov replied:
LDFLAGS_FOR_TESTDLL could be used to add required flags

thiru-mcw replied:
Adding `-L$(TOOLCHAIN_PATH)/$(target)/lib/w32api/` to `LDFLAGS_FOR_TESTDLL` can successfully build the `devdsp.exe` test case for `AArch64-pc-cygwin` target. 

However, I Observed this particular change is not required for the `x86_64-pc-cygwin` target.

eukarpov replied:
Then it should be configured only for aarch64.

Blackhex replied:
I'd say that the proper solution to the issue with missing `libwinmm` is https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/pull/301/files#diff-0721c1092ad60268b58ab59a49b4d991f826f8f004ffbf3661f9b754f1fce58fR94. The current solution - referencing the library from `mingw-w64-crt` build - does not seem right to me. The Cygwin build should not depend on MinGW build folder.

thiru-mcw replied:
Done, Thanks for the Inputs. 
Added the same change in cyg-test2 branch. Tested locally

Blackhex commented on winsup/testsuite/winsup.api/pthread/cancel5.c:
```diff
@@ -78,6 +78,15 @@ mythread(void * arg)
    * We wait up to 30 seconds for a cancelation to be applied to us.
    */
+
+#if defined(__x86_64__) || defined(__i386__)  // Check for x86 architectures
```
Since this is used on two places already woun't there be an appropriate header where it can be defined?

thiru-mcw replied:
Yes, there is a header `test.h` in the same directory, and adding it there works.

However, this header is also used in 72 other tests, so I created a new header, `cpu_relax.h`, and added the definition there. 

Is this fine, or should I go ahead and modify `test.h` instead? Like to hear your inputs on this @Blackhex

Blackhex replied:
Fine for me, thank you for the update.

Blackhex reviewed (commented):
Overall looks good. I can approve once https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/pull/12/files#r1998101812 comment will be answsered or addressed.

eukarpov commented on winsup/Makefile.am:
```diff
@@ -14,10 +14,10 @@ cygdoc_DATA = \
 endif
 
-cygserver: cygwin
+cygserver testsuite: cygwin
```
How does it work on x64?

thiru-mcw replied:
The `test suite-related` changes from commit bb23c9b2c529274a0d468a2ba33e04901a3ab134, which provided a workaround for building the Cygwin cross-compilation toolchain through stage1 bootstrapping, have been reverted. 

Additionally, the changes related to `utils` can also be reverted, as they do not impact the toolchain build and still allow successful compilation of "Hello World" executables.

Reference: efa39b035e60022328f5eb91dbf52854547d9ca6

eukarpov commented on winsup/testsuite/Makefile.am:
```diff
@@ -322,15 +321,14 @@ LDADD_FOR_TESTDLL = $(builddir)/../cygwin/libcygwin.a -lgcc -lkernel32 -luser32
 # all tests
-TESTS = $(check_PROGRAMS) \
-	mingw/cygload
+TESTS = $(check_PROGRAMS)
```
Could you please clarify this change?

thiru-mcw replied:
The `mingw/cygload` was removed due to missing Makefile rules (`No rule to make target 'mingw/cygload'`). 
This has been identified as a deficiency in the current toolchain build system.

eukarpov replied:
Thanks for the clarification. Please add a comment.

thiru-mcw replied:
This particular test requires MinGW compiler to be built, so this is temporarily excluded.

https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/7a762b853e1a994616adab1e499c866629be0e74/winsup/testsuite/mingw/Makefile.am#L11-L13

eukarpov commented on winsup/testsuite/winsup.api/ltp/alarm01.c:
```diff
@@ -188,7 +188,7 @@ main(int ac, char **av)
 setup()
 {
-    void trapper();
+    void trapper(int sig);
```
I guess starting from this change and many other changes below  
are not needed after configuring the build with gnu89 support. Which is not enabled by default in the aarch64-pc-cygwin toolchain.

Blackhex replied:
Cygwin community is open to modernize the codebase. We've already upstreamed two patches doing that on the functional codebase, the update of testing codebase can follow. Though from puristic point of view, it would be better if this testing code modernization would be a separate PR preceeding this one.

eukarpov replied:
Not sure why it is needed if it can be fixed by one config change.

thiru-mcw replied:
@eukarpov , As you suggested most of the changes in this PR can be avoided by adding the following flags,
```
AM_CFLAGS = -std=gnu89  
AM_CXXFLAGS = -std=gnu++98  
```
With this approach, only a few changes, such as `CPU_RELAX` and `libwinmm` modifications, would remain in the PR.  

> *Cygwin community is open to modernizing the codebase.*  

Based on upstream commits [[Ref1](https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/commit/c79954439e6d91dc6f421a761953f0d935c5f3d5), [Ref2](https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/commit/d636b11d2e167cbcc310000ac1a977219b819fee)] I proceeded with these fixes to ensure compatibility with modern C standards.  
Please confirm which approach would work best for this PR.

Blackhex replied:
I am fine with both if follow up PR will be issued.

eukarpov replied:
Please move the standard migration out of this PR and the main branch will be rebased once it is upstreamed. Still not sure why it is needed.

thiru-mcw replied:
> Please move the standard migration out of this PR

Sure, Planning to submit changes related to codebase modernization directly to cygwin mailing list. Once done and merged will amend this PR.

thiru-mcw replied:
> AM_CFLAGS = -std=gnu89 

Is added to the make file, this makes this PR with minimal changes.
Thanks @eukarpov

eukarpov reviewed (commented):
I would expect minimal changes such as CPU_RELAX, other changes if needed should be explained in comments why it is different from x64 and how it might be addressed in the future.

thiru-mcw commented:
> I would expect minimal changes such as CPU_RELAX, other changes if needed should be explained in comments why it is different from x64 and how it might be addressed in the future.

The `cpu-relax.h` header is required to handle architecture-specific instructions in current implementation of `cancel5.c` and `cancel3.c` which uses the `PAUSE` instruction, that is specific to the x86_64 :  


https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/7a762b853e1a994616adab1e499c866629be0e74/winsup/testsuite/winsup.api/pthread/cancel3.c#L91

For aarch64, an equivalent instruction is `YIELD`, as inspired from [Ref](https://postgrespro.com/list/thread-id/2487147) and ARM's official [Doc](https://developer.arm.com/documentation/100076/0100/A64-Instruction-Set-Reference/A64-General-Instructions/YIELD).  By introducing `cpu-relax.h`, we can ensure compatibility across platforms.

eukarpov commented on winsup/testsuite/Makefile.am:
```diff
@@ -49,7 +49,6 @@ check_PROGRAMS = \
 	winsup.api/sigchld \
 	winsup.api/signal-into-win32-api \
 	winsup.api/systemcall \
-	winsup.api/user_malloc \
```
what about this change?

thiru-mcw replied:
As mentioned here https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/pull/12#issuecomment-2724138829 building the `user_malloc` test leads to relocation issues on ARM64.

```
  CC       winsup.api/user_malloc.o
  CCLD     winsup.api/user_malloc.exe
./../cygwin/libcygwin.a(t-cygwin1_dll_d000024.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp___errno' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000024.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000037.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp___getreent' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000037.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000048.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp___main' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000048.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000132.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp_access' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000132.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000425.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp_dup' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000425.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000464.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp_execl' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000464.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000471.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp_exit' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000471.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000496.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp_fclose' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000496.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000507.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp_fdopen' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000507.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000530.o):(.text+0x0): relocation truncated to fit: IMAGE_REL_ARM64_REL21 against symbol `__imp_fflush' defined in .idata$5 section in ./../cygwin/libcygwin.a(cygwin1_dll_d000530.o)
./../cygwin/libcygwin.a(t-cygwin1_dll_d000545.o):(.text+0x0): additional relocation overflows omitted from the output
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4328: winsup.api/user_malloc.exe] Error 1"
```

Blackhex reviewed (changes requested):
Unfortunatelly, mi biggest concern with `winmm` linking wasn't addressed. Otherwise, it's OK for me.

Blackhex reviewed (approved):
Lood good for me as a first step towards making the tests passing.

thiru-mcw commented:
[[PATCH] Cygwin: Fix compatibility with GCC 15](https://cygwin.com/pipermail/cygwin-patches/2025q1/013561.html)

[Cygwin: testsuite: Fix compilation for arm64](https://github.com/cygwin/cygwin/commit/6c5537c0298e9135ff27119c6a18dcc6f94459f5)
Introduce a cpu_relax.h header to provide architecture-specific
processor idling instructions for the `cancel3` and `cancel5` tests.


[Cygwin: testsuite: Fix compatibility with GCC 15](https://github.com/cygwin/cygwin/commit/cce2ffd374e2ab4507cb973c74348cd1be4d106e)
GCC 15 defaults to `-std=gnu23`, causing build failures in `testsuite` due to outdated C function declarations. This patch updates the function declarations to align with modern standards.

- As the above PRs are merged the scope of this PR is reduced. 
- `AM_CFLAGS = -std=gnu89` is added to bypass only 2 tests (symlink, sbrk) for which a separate patch is submitted to upstream repository [[under review](https://cygwin.com/pipermail/cygwin-patches/2025q2/013605.html)]

eukarpov commented:
Does the PR pass validation on CI with current changes?

Blackhex commented:
@thiru-mcw Why the https://github.com/cygwin/cygwin/commit/6c5537c0298e9135ff27119c6a18dcc6f94459f5 commit was removed from this branch? Was it already unstreamed? Currently the build of tests fails with:
```
cygwin/winsup/testsuite/winsup.api/pthread/cancel3.c:91: Error: unknown mnemonic `pause' -- `pause'
```
on this branch.

thiru-mcw commented:
> @thiru-mcw Why the [cygwin/cygwin@6c5537c](https://github.com/cygwin/cygwin/commit/6c5537c0298e9135ff27119c6a18dcc6f94459f5) commit was removed from this branch? Was it already unstreamed? Currently the build of tests fails with:
> 
> ```
> cygwin/winsup/testsuite/winsup.api/pthread/cancel3.c:91: Error: unknown mnemonic `pause' -- `pause'
> ```
> 
> on this branch.

Yes, Radek I have removed those changes from this branch as they are already upstreamed and accepted by cygwin maintainers.
[REF-1](https://github.com/cygwin/cygwin/commit/6c5537c0298e9135ff27119c6a18dcc6f94459f5)
[REF-2](https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/pull/12#issuecomment-2772444685)

These commits can be brought directly to `woarm64` branch once rebased.

eukarpov commented:
I can improve this PR once it passes validation.

Blackhex commented:
Under the current circumstances that the `woarm64` is not rebased on upstream, you can consider [this PR's](https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/pull/301) [workflow run](https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/14221087728) as a validation of this PR. The `pause` changes have been cherry-picked to the branch based on this PR's one.

eukarpov commented:
The PR can be merged only after rebase, otherwise it might break current validations and builds.

thiru-mcw commented on winsup/testsuite/Makefile.am:
```diff
@@ -321,6 +323,7 @@ LDFLAGS_FOR_TESTDLL = -nodefaultlibs
 LDADD_FOR_TESTDLL = $(builddir)/../cygwin/libcygwin.a -lgcc -lkernel32 -luser32
 
 # flags for test executables
+AM_CFLAGS = -std=gnu89  
```
I believe this change can be reverted, as all the commits related to test-fixes has landed in the master of the cygwin-upstream repo.

Ref: [test-fix commits](https://github.com/cygwin/cygwin/commits/main/?author=thiru-mcw)

That can be brought here to woarm64 by a rebase.

thiru-mcw committed:
Enable winsup test suite compilation

User_malloc is temporarily removed due to compilation errors, workaround to
allow the testsuite to reach completion and display the summary.
mingw/cygload test is excluded due to the toolchain limitations.
Removing cygdrop $cygrun to make the tests pass while testing on wsl-env.

Blackhex committed:
Workaround for undefined reference to `_environ`

Blackhex commented:
@vejbomar @thiru-mcw @eukarpov Please, review. Updated and workarounds for https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/issues/258 were added. Validated by https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/15139973579

thiru-mcw commented:
> @thiru-mcw Please, review. 

LGTM..
